lib/Doctrine/Common/Collections/ReadableCollection.php

```diff
@@ -0,0 +1,209 @@
+     *               elements in the collection.
+     * @psalm-return TKey[]
+     */
+    public function getKeys();
+
+    /**
+     * Gets all values of the collection.
+     *
+     * @return mixed[] The values of all elements in the collection, in the
+     *                 order they appear in the collection.
+     * @psalm-return list<T>
+     */
+    public function getValues();
+
+    /**
+     * Gets a native PHP array representation of the collection.
+     *
+     * @return mixed[]
+     * @psalm-return array<TKey,T>
+     */
+    public function toArray();
+
+    /**
+     * Sets the internal iterator to the first element in the collection and returns this element.
+     *
+     * @return mixed
+     * @psalm-return T|false
+     */
+    public function first();
+
+    /**
+     * Sets the internal iterator to the last element in the collection and returns this element.
+     *
+     * @return mixed
+     * @psalm-return T|false
+     */
+    public function last();
+
+    /**
+     * Gets the key/index of the element at the current iterator position.
+     *
+     * @return int|string|null
+     * @psalm-return TKey|null
+     */
+    public function key();
+
+    /**
+     * Gets the element of the collection at the current iterator position.
+     *
+     * @return mixed
+     * @psalm-return T|false
+     */
+    public function current();
+
+    /**
+     * Moves the internal iterator position to the next element and returns this element.
+     *
+     * @return mixed
+     * @psalm-return T|false
+     */
+    public function next();
+
+    /**
+     * Tests for the existence of an element that satisfies the given predicate.
+     *
+     * @param Closure $p The predicate.
+     * @psalm-param Closure(TKey=, T=):bool $p
+     *
+     * @return bool TRUE if the predicate is TRUE for at least one element, FALSE otherwise.
+     */
+    public function exists(Closure $p);
+
+    /**
+     * Returns all the elements of this collection that satisfy the predicate p.
+     * The order of the elements is preserved.
+     *
+     * @param Closure $p The predicate used for filtering.
+     * @psalm-param Closure(T=):bool $p
+     *
+     * @return Collection<mixed> A collection with the results of the filter operation.
+     * @psalm-return Collection<TKey, T>
+     */
+    public function filter(Closure $p);
+
+    /**
+     * Tests whether the given predicate p holds for all elements of this collection.
+     *
+     * @param Closure $p The predicate.
+     * @psalm-param Closure(TKey=, T=):bool $p
+     *
+     * @return bool TRUE, if the predicate yields TRUE for all elements, FALSE otherwise.
+     */
+    public function forAll(Closure $p);
+
+    /**
+     * Applies the given function to each element in the collection and returns
+     * a new collection with the elements returned by the function.
+     *
+     * @psalm-param Closure(T=):U $func
```

Comment:
That was introduced in 0c8bdd12640ee68286baf61fb5cae37b90c3af1a

@derrabus , do you remember why?

Reply:
This was introduced in 744afc199626c1638113da4cd4bf2222c91dcd13

Reply:
`Closure(T=):U` means that the closure could be called with no argument, which then forces all consumers of the interface to write their callbacks with optional arguments. This does not make any sense.

On the other hand, given that PHP does not forbid passing extra arguments to a callback, it is totally fine to pass a callback expecting no argument to `Closure(T):U` (it will simply ignore the provided argument).

So `Closure(T):U` is the right type here. We want our code calling `$func` to always pass 1 argument.

Reply:
Ok, I'll contribute that to 1.7.x